### PR TITLE
Relax version constraint for Ruby HEAD testing

### PR DESCRIPTION
**Pull request by lloeki**
### Problem

```
  Bundler found conflicting requirements for the Ruby version:
    In Gemfile:
      Ruby
  
      ddtrace was resolved to 0.54.1, which depends on
        Ruby (>= 2.1.0, < 3.2)
  
  Ruby (>= 2.1.0, < 3.2), which is required by gem 'ddtrace', is not available in
```

The reason for this version constraint is because `ruby2_keywords` is expected to disappear on 3.2 but the current code relies on it.

### Suggestions by @ivoanjo 

- Disable testing of ruby-head in CI
- Bump maximum Ruby version to 3.3
- Other?

### Opinions by @lloeki 

> Disable testing of ruby-head in CI

Not too fond of this, maybe make it as "allow failure" (if that's not already the case) so as not to block dependent jobs, but GHA is terrible and the UI would mark the whole build as failure even though it's OK (I wish it had a "warning" visual state as in GitLab CI)

> Bump maximum Ruby version to 3.3

I'd be OK with that for CI, I mean it's consistent to allow for the final release version when you also allow for dev/preview. But it's also kind of misleading to customers.

> Other?

Keep it in source control as what it should be for a public release + add a step to the HEAD job that either seds to 3.3 or outright removes the `spec.required_ruby_version` line.

### Cause

Although this version constraint should work ([doc](https://guides.rubygems.org/specification-reference/#required_ruby_version=), tested on 3.1 instead of 3.2 for ease of docker image pulling, but that's largely inconsequential)

```
$ ruby -e 'p Gem::Requirement.new("< 3.1.0").satisfied_by?(Gem.ruby_version)'
true # on 3.1.0-preview1
false # on 3.1.0
```

(for the sake of being thorough, appending .a to the requirement is how you're supposed to exclude prereleases:)

```
$ ruby -e 'p Gem::Requirement.new("< 3.1.0.a").satisfied_by?(Gem.ruby_version)'
false # on both 3.1.0-preview1 and 3.1.0
```

... it appears it does not: currently the following fails on both `3.1.0` and `3.1.0-preview1`, which baffles me.

```
# Gemfile
source 'https://rubygems.org'

gemspec
```

and:

```
# foo.gemspec
Gem::Specification.new do |spec|
  spec.name                  = 'foo'
  spec.version               = '0.0.1'
  spec.required_ruby_version = '< 3.1.0'
  spec.required_rubygems_version = '>= 2.0.0'
  spec.summary = 'foo'
  spec.authors = ['foo']
  spec.files = ['foo.rb']
end
```

### CODE REVIEW

**Comment by lloeki** [comment]
One test fails:

```
An error occurred while loading spec_helper.
Failure/Error: require 'pry'

NameError:
  undefined method `=~' for class `Pry::Code'
# ./vendor/bundle/ruby/3.2.0/gems/pry-0.13.1/lib/pry/code.rb:342:in `<class:Code>'
# ./vendor/bundle/ruby/3.2.0/gems/pry-0.13.1/lib/pry/code.rb:32:in `<class:Pry>'
# ./vendor/bundle/ruby/3.2.0/gems/pry-0.13.1/lib/pry/code.rb:5:in `<top (required)>'
# ./vendor/bundle/ruby/3.2.0/gems/pry-0.13.1/lib/pry.rb:70:in `require'
# ./vendor/bundle/ruby/3.2.0/gems/pry-0.13.1/lib/pry.rb:70:in `<top (required)>'
# ./spec/spec_helper.rb:4:in `require'
# ./spec/spec_helper.rb:4:in `<top (required)>'
```

I venture this is related to `pry` needing an update for 3.2.

File: .github/workflows/test-head.yaml [hunk 1]
```diff
@@ -12,6 +12,7 @@ jobs:
       JRUBY_OPTS: --dev
     steps:
       - uses: actions/checkout@v2
+      - run: sed -i~ -e '/spec\.required_ruby_version/d' ddtrace.gemspec
```
Comment (ivoanjo) [inline .github/workflows/test-head.yaml:15]: Minor:

* Would it make sense to add a small comment saying something like "Skip gem maximum version restriction when testing with ruby-head", just to add a bit of context to this magic incantation?
* My sed-foo is not amazing; I couldn't figure out what the `-i~` does and how differs from `-i`

**Review by ivoanjo — APPROVED** [review]
Thanks for looking into this!

I noticed that the tests are still broken because:

* pry 0.13 is broken on Ruby 3.2; but pry 0.14 works
* pry 0.14 is not picked up because the latest pry-byebug is restricted to pry 0.13

TL;DR this fixed it for me:

```diff
diff --git a/Gemfile b/Gemfile
index a70835653..2ec8b5d41 100644
--- a/Gemfile
+++ b/Gemfile
@@ -17,7 +17,7 @@ gem 'pimpmychangelog', '>= 0.1.2'
 gem 'pry'
 if RUBY_PLATFORM != 'java'
   # There's a few incompatibilities between pry/pry-byebug on older Rubies
-  gem 'pry-byebug' if RUBY_VERSION >= '2.6.0' && RUBY_ENGINE != 'truffleruby'
+  gem 'pry-byebug' if RUBY_VERSION >= '2.6.0' && RUBY_ENGINE != 'truffleruby' && RUBY_VERSION < '3.2.0'
   gem 'pry-nav' if RUBY_VERSION < '2.6.0'
   gem 'pry-stack_explorer'
 else
```

I'm guessing the pry-byebug folks will at some point add support for pry 0.14 :)

File: .github/workflows/test-head.yaml (hunk 1 above)
Comment (lloeki) [reply .github/workflows/test-head.yaml:15]: `man sed` ;)

`-i` means "edit in place", and takes a value to append to the filename to backup the original. On GNU systems `-i` has this value optional, which means discard the original, but on BSD systems it's not, so the value would be `-e`. Coincidentally, `-e` is optional on both when there's only one pattern, so on BSD systems `sed -i -e 'pattern' file` works but silently breaks the user expectation by saving the original as `file-e`! So, that's a portable way to write inline replacements with sed. As for appending `~`, it's a historical convention not entirely unlike `.bak`.

Comment (lloeki) [reply .github/workflows/test-head.yaml:15]: > "Skip gem maximum version restriction when testing with ruby-head"

Thought about it but given that it's in the `test-head.yaml file` (and on the other one there's a `if:` clause testing for `head`) I don't find this comment to be adding value (it sounds like a "what")

That said, mentioning the "why" context (being unable to include prereleases yet exclude releases) sounds like a good idea.

Comment (ivoanjo) [reply .github/workflows/test-head.yaml:15]: Thanks for explaining the sed magic!

Also agree with your note on the comment -- I was thinking from the POV that we sometimes copy pasta CI stuff and add more versions and whatnot so the context of "this sed only makes sense for head" may be lost, but the better fix is definitely your suggestion of explaining why :)

**Review by marcotc — APPROVED** [review]
Thank you, @lloeki!

**Comment by ivoanjo** [comment]
Thanks @lloeki, going to merge this in so that our CI can be a beautiful green again :)
